llama-index-integrations/llms/llama-index-llms-githubllm/llama_index/llms/githubllm/base.py

```diff
@@ -0,0 +1,296 @@
+
+        # Multi-turn chat
+        messages = [
+            ChatMessage(role="user", content="Tell me about the French Revolution."),
+            ChatMessage(role="assistant", content="The French Revolution was a period of major social and political upheaval in France that began in 1789 with the Storming of the Bastille and ended in the late 1790s with the ascent of Napoleon Bonaparte. It was partially carried forward by Napoleon during the later expansion of the French Empire. The Revolution overthrew the monarchy, established a republic, catalyzed violent periods of political turmoil, and fundamentally altered French history."),
+            ChatMessage(role="user", content="What were the main causes?")
+        ]
+
+        response = llm.chat(messages)
+        print(response)
+
+        # Streaming chat
+        for chunk in llm.stream_chat([ChatMessage(role="user", content="Can you elaborate on the Reign of Terror?")]):
+            print(chunk.message.content, end='', flush=True)
+    """
+
+    github_endpoint_url: str = Field(
+        default="https://models.inference.ai.azure.com/chat/completions"
+    )
+    model: str = Field(description="The model to use for inference.")
+    system_prompt: str = Field(default="You are a helpful assistant.")
+    use_azure_fallback: bool = Field(default=True)
+    temperature: float = Field(default=0.7)
+    max_tokens: int = Field(default=None)
+
+    _rate_limit_reset_time: float = PrivateAttr(default=0)
+    _request_count: int = PrivateAttr(default=0)
+    _max_requests_per_minute: int = PrivateAttr(default=15)  # Adjust based on your tier
+    _max_requests_per_day: int = PrivateAttr(default=150)  # Adjust based on your tier
+
+    SUPPORTED_MODELS = [
+        "AI21-Jamba-Instruct",
+        "cohere-command-r",
+        "cohere-command-r-plus",
+        "cohere-embed-v3-english",
+        "cohere-embed-v3-multilingual",
+        "meta-llama-3-70b-instruct",
+        "meta-llama-3-8b-instruct",
+        "meta-llama-3.1-405b-instruct",
+        "meta-llama-3.1-70b-instruct",
+        "meta-llama-3.1-8b-instruct",
+        "mistral-large",
+        "mistral-large-2407",
+        "mistral-nemo",
+        "mistral-small",
+        "gpt-4o",
+        "gpt-4o-mini",
+        "phi-3-medium-instruct-128k",
+        "phi-3-medium-instruct-4k",
+        "phi-3-mini-instruct-128k",
+        "phi-3-mini-instruct-4k",
+        "phi-3-small-instruct-128k",
+        "phi-3-small-instruct-8k",
+    ]
+
+    @validator("model")
+    def validate_model(cls, v):
+        if v.lower() not in [model.lower() for model in cls.SUPPORTED_MODELS]:
+            raise ValueError(
+                f"Model {v} is not supported. Please choose from {cls.SUPPORTED_MODELS}"
+            )
+        return v
+
+    @property
+    def metadata(self) -> LLMMetadata:
+        """Get LLM metadata."""
+        return LLMMetadata(
+            context_window=4096,  # Assuming a default context window, adjust as needed
```

Comment:
Do we want to hardcode this, or should it be passed in?

Reply:
i could perhaps have the 22 models' context windows vary according to user input and a sanity check ?

Reply:
Hardcoding the context windows per model is a valid strat (most llm classes do this)